File: src/main/java/com/refinedmods/refinedstorage/network/grid/GridFluidInsertHeldMessage.java
package com.refinedmods.refinedstorage.network.grid;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.api.network.grid.IGrid;
import com.refinedmods.refinedstorage.container.GridContainerMenu;

import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.neoforged.neoforge.network.handling.PlayPayloadContext;

public class GridFluidInsertHeldMessage implements CustomPacketPayload {
    public static final ResourceLocation ID = new ResourceLocation(RS.ID, "grid_fluid_insert_held");

    public static void handle(GridFluidInsertHeldMessage msg, PlayPayloadContext ctx) {
        ctx.player().ifPresent(player -> ctx.workHandler().submitAsync(() -> {
            AbstractContainerMenu container = player.containerMenu;

            if (container instanceof GridContainerMenu) {
                IGrid grid = ((GridContainerMenu) container).getGrid();

                if (grid.getFluidHandler() != null) {
                    grid.getFluidHandler().onInsertHeldContainer((ServerPlayer) player);
                }
            }
        }));
    }

    @Override
    public void write(FriendlyByteBuf buf) {
        // no op
    }

    @Override
    public ResourceLocation id() {
        return ID;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/network/grid/GridFluidInsertHeldMessage.kt
package com.refinedmods.refinedstorage.network.grid

import com.refinedmods.refinedstorage.container.GridContainer
import net.minecraft.entity.player.ServerPlayerEntity
import net.minecraft.inventory.container.Container
import net.minecraft.network.PacketByteBuf
import net.minecraftforge.fml.network.NetworkEvent
import java.util.function.Supplier

object GridFluidInsertHeldMessage {
    fun decode(buf: PacketByteBuf?): GridFluidInsertHeldMessage {
        return GridFluidInsertHeldMessage()
    }

    fun encode(message: GridFluidInsertHeldMessage?, buf: PacketByteBuf?) {}
    fun handle(message: GridFluidInsertHeldMessage?, ctx: Supplier<NetworkEvent.Context>) {
        val player: ServerPlayerEntity = ctx.get().getSender()
        if (player != null) {
            ctx.get().enqueueWork({
                val container: Container = player.openContainer
                if (container is GridContainer) {
                    val grid = (container as GridContainer).grid
                    if (grid!!.fluidHandler != null) {
                        grid.fluidHandler!!.onInsertHeldContainer(player)
                    }
                }
            })
        }
        ctx.get().setPacketHandled(true)
    }
}